Theme Showcase: Do not show wporg themes if they are also retired wpcom themes

taipeicoder opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #78091

## Proposed Changes

The current implementation shows a wporg theme even if its wpcom counterpart is retired. For instance, the theme Maxwell exists in both wporg and wpcom, but it's retired in wpcom. This leads to a scenario where the Theme Showcase will show the wporg version of the theme, but when the user clicks on the theme, the detail page will say that the theme is retired. This behavior is confusing, and might lead to issues as described in #78091.

This PR updates the logic so that retired wpcom themes are always omitted from the Theme Showcase, even if their wporg counterparts exist. In other words, when users search for Maxwell, even though the theme exists in wporg, it will not be shown in the Theme Showcase listing because it's retired in wpcom.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Head to the Theme Showcase.
* Search for a retired theme that also exists in wporg, such as Maxwell or Adelle.
* Ensure that the theme is not shown in the theme listing.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

taipeicoder committed:
Add theme is retired check for theme card

arthur791004 commented:
Tested but the Maxwell theme is `null` so the activate action is still there

![image](https://github.com/Automattic/wp-calypso/assets/13596067/37446de9-ccd7-4066-a959-827a93168033)

arthur791004 commented on client/state/themes/selectors/is-theme-retired.js:
```diff
@@ -0,0 +1,15 @@
+ * @returns {boolean}        True if the theme is retired
+ */
+export function isThemeRetired( state, themeId ) {
+	return isRetired( getTheme( state, 'wpcom', themeId ) );
```
The theme might not be fetched. So, one option is to get theme from wporg as well and add the slug to `RETIRED_THEME_SLUGS_SET`

taipeicoder replied:
Hmmm the `retired` property is only from wpcom, which is why we only query for the theme in wpcom. How can I reproduce the theme being `null` 🤔 ?

arthur791004 replied:
The theme was null when I went to the Theme Showcase and searched “Maxwell”. Or maybe you can head to the search result page directly 🤔

taipeicoder replied:
@arthur791004 with D113775-code merged, the query will no longer return `null`. Could you test again? Thanks!

arthur791004 replied:
We have to pass `retired=true` to themes endpoint to get the result

taipeicoder replied:
Oops, I forgot to commit 😅

arthur791004 reviewed (approved):
Tested again and looks good to me 👍

arthur791004 commented:
BTW, D113775-code has a side effect on the search result...we have to hide the retired themes from wpcom on client side as well

![image](https://github.com/Automattic/wp-calypso/assets/13596067/942ac734-0b96-4e38-856b-82661e96d223)

taipeicoder commented:
> BTW, D113775-code has a side effect on the search result...we have to hide the retired themes from wpcom on client side as well
> 
> ![image](https://user-images.githubusercontent.com/13596067/246388919-942ac734-0b96-4e38-856b-82661e96d223.png)

Was your search term "blank canvas". I could only see the one in wpcom 🤔

![Screenshot 2023-06-16 at 7 00 44 PM](https://github.com/Automattic/wp-calypso/assets/797888/87a4df2b-f501-4e01-8c9a-7524ee2906ff)

Were you using the calypso.live link?

arthur791004 commented:
> Were you using the calypso.live link?

Yes, I was using the calypso.live link and searched for the “blank canvas”. And here is the result on production:

![image](https://github.com/Automattic/wp-calypso/assets/13596067/946e286b-77e5-4b54-ac62-cbec90621f02)

taipeicoder committed:
Linting

taipeicoder committed:
Revert unnecessary commits

taipeicoder committed:
Revert unnecessary commits

taipeicoder commented:
@arthur791004 I've changed the PR proposal to an approach that I think makes more sense, and updated the description accordingly. I've moved the PR to Needs Review again 🙂

arthur791004 commented on client/my-sites/themes/helpers.js:
```diff
@@ -162,7 +163,12 @@ export function interlaceThemes( wpComThemes, wpOrgThemes, searchTerm, isLastPag
+	// The themes endpoint returns retired themes when search term exists.
+	// See: https://github.com/Automattic/wp-calypso/pull/78231
+	interlacedThemes.push(
+		...( searchTerm ? restWpComThemes.filter( ( theme ) => ! theme.retired ) : restWpComThemes )
```
[nit] We can filter out the retired themes regardless of the search term. But the current way is also fine 👍

taipeicoder replied:
I'll leave the logic in for now, since I like the idea of the small performance gain! 🙂

arthur791004 reviewed (approved):
Tested again and looks good 🎉

![image](https://github.com/Automattic/wp-calypso/assets/13596067/6c32baaf-3846-4d79-944e-b236cc085c39)

![image](https://github.com/Automattic/wp-calypso/assets/13596067/887fb423-00ed-4206-942e-5537100a004f)
